### PR TITLE
ICU-13565 Break Iteration, remove the dictionary bit from the impleme…

### PR DESCRIPTION
**Pull request by aheninger**
…ntation.

For identifying text that needs to be handled by a word dictionary for Break Iteration,
change from using a bit in the character category to sorting all dictionary categories
together, and recording the boundary between the non-dictionary and dictionary ranges.

This is internal to the implementation. It does not affect behavior.
It does increase the number of character categories that can be handled using a
compact 8 bit Trie, from 127 to 255.

##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-13565
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/common/rbbisetb.cpp [hunk 1]
```diff
@@ -255,24 +276,16 @@ void RBBISetBuilder::buildRanges() {
...
                         fStatus);
 
-    bool use8Bits = getNumCharCategories() <= kMaxCharCategoriesFor8BitsTrie;
-    for (rlRange = fRangeList; rlRange!=0 && U_SUCCESS(*fStatus); rlRange=rlRange->fNext) {
-        uint32_t value = rlRange->fNum;
-        if (use8Bits && ((value & RuleBasedBreakIterator::kDictBit) != 0)) {
-            U_ASSERT((value & RuleBasedBreakIterator::kDictBitFor8BitsTrie) == 0);
-            value = RuleBasedBreakIterator::kDictBitFor8BitsTrie | (value & ~RuleBasedBreakIterator::kDictBit);
-        }
+    for (RangeDescriptor *range = fRangeList; range!=0 && U_SUCCESS(*fStatus); range=range->fNext) {
```
Comment (markusicu) [inline icu4c/source/common/rbbisetb.cpp:284]: optional: `range != nullptr`

File: icu4c/source/common/rbbisetb.cpp [hunk 2]
```diff
@@ -494,7 +517,7 @@ void RBBISetBuilder::printRangeGroups() {
             }
 
             i = 0;
-            for (tRange = rlRange; tRange != 0; tRange = tRange->fNext) {
+            for (RangeDescriptor *tRange = rlRange; tRange != 0; tRange = tRange->fNext) {
```
Comment (markusicu) [inline icu4c/source/common/rbbisetb.cpp:520]: optional: `tRange != nullptr`

File: icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java [hunk 3]
```diff
@@ -239,25 +231,49 @@ void buildRanges() {
...
                 }
             }
             if (rlRange.fNum == 0) {
-                fGroupCount ++;
-                rlRange.fNum = fGroupCount+2;
-                rlRange.setDictionaryFlag();
-                addValToSets(rlRange.fIncludesSets, fGroupCount+2);
+                rlRange.fFirstInGroup = true;
+                if (rlRange.isDictionaryRange()) {
+                    rlRange.fNum = ++dictGroupCount;
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java:251]: see questions in C++

File: icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java [hunk 4]
```diff
@@ -324,16 +330,20 @@ void buildTrie() {
...
+            int rangeNum = rd.fNum;
             if (rangeNum == categories.second) {
-                rd.fNum = categories.first | rangeDict;
+                rd.fNum = categories.first;
             } else if (rangeNum > categories.second) {
                 rd.fNum--;
             }
         }
         --fGroupCount;
+        if (fDictCategoriesStart >= categories.second) {
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java:344]: optional: turn around as suggested in C++

File: icu4j/main/classes/core/src/com/ibm/icu/text/RuleBasedBreakIterator.java [hunk 5]
```diff
@@ -1222,9 +1209,10 @@ void populateDictionary(int startPos, int endPos,
             fText.setIndex(rangeStart);
             int     c = CharacterIteration.current32(fText);
             category = (short)fRData.fTrie.get(c);
+            int dictCategoryStart = fRData.fFTable.fDictCategoriesStart;
```
Comment (markusicu) [inline icu4j/main/classes/core/src/com/ibm/icu/text/RuleBasedBreakIterator.java:1212]: optional: shorter dictStart ?

File: icu4c/source/common/rbbisetb.cpp [hunk 6]
```diff
@@ -224,11 +247,9 @@ void RBBISetBuilder::buildRanges() {
     //     subtree for each UnicodeSet that contains the string {eof}
     //   Because {bof} and {eof} are not a characters in the normal sense,
```
[the remark this answers is not in the document]
Comment (aheninger) [reply icu4c/source/common/rbbisetb.cpp]: Done

File: icu4c/source/common/rbbisetb.cpp (hunk 1 above)
Comment (aheninger) [reply icu4c/source/common/rbbisetb.cpp]: Done

File: icu4c/source/common/rbbisetb.cpp [hunk 7]
```diff
@@ -281,16 +294,21 @@ void RBBISetBuilder::buildTrie() {
...
+        int32_t rangeNum = rd->fNum;
         if (rangeNum == categories.second) {
-            rd->fNum = categories.first | rangeDict;
+            rd->fNum = categories.first;
         } else if (rangeNum > categories.second) {
             rd->fNum--;
         }
     }
     --fGroupCount;
+    if (fDictCategoriesStart >= categories.second) {
```
[the remark this answers is not in the document]
Comment (aheninger) [reply icu4c/source/common/rbbisetb.cpp]: Done

File: icu4c/source/common/rbbisetb.cpp (hunk 2 above)
Comment (aheninger) [reply icu4c/source/common/rbbisetb.cpp]: Done

File: icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java (hunk 3 above)
Comment (aheninger) [reply icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java:251]: And answered in C++.

File: icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java (hunk 4 above)
Comment (aheninger) [reply icu4j/main/classes/core/src/com/ibm/icu/text/RBBISetBuilder.java]: Done

File: icu4j/main/classes/core/src/com/ibm/icu/text/RuleBasedBreakIterator.java (hunk 5 above)
Comment (aheninger) [reply icu4j/main/classes/core/src/com/ibm/icu/text/RuleBasedBreakIterator.java]: Done

**Pull request merged**